File: lib/stats.js
'use strict';

/*
 * Snapshot analytics: aggregate composition of the loaded lists plus the most
 * recently designated parties.
 *
 * Everything here is derived from the SAME immutable snapshot the screening
 * endpoint uses, so a statistic and a search result can never disagree. The
 * result is memoized per snapshot object (a refresh swaps the object and
 * invalidates it) — a full pass over ~38k entities is ~100ms and must not run
 * on every request.
 *
 * Dates: each authority publishes its listing date in its own format — OFAC/EU/
 * UN/BIS in ISO, UK OFSI in DD/MM/YYYY. They are normalized to ISO here so the
 * timeline and the "recent" ranking are comparable across authorities. The date
 * is the AUTHORITY'S designation/publication date, not when we ingested it.
 */

const { canonicalCountry, parseCountries } = require('./countries');

const RECENT_LIMIT = 150;        // most recent designations across all lists
const RECENT_PER_AUTHORITY = 40; // …plus this many per authority, so the
                                 // authority filter is never empty for a list
                                 // that has not published in the global window
const TOP_N = 15;           // rows per "top" breakdown
const TIMELINE_YEARS = 16;
const TIMELINE_MONTHS = 24;

const DAY = 86400000;

// Accepts "YYYY-MM-DD" / ISO timestamps and "D/M/YYYY" (UK OFSI). Returns
// "YYYY-MM-DD", or '' when the value is missing or not a date we can trust.
function normalizeDate(v) {
  if (!v) return '';
  const s = String(v).trim();
  let m = s.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (m) return `${m[1]}-${m[2]}-${m[3]}`;
  m = s.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/); // day/month/year
  if (m) {
    const d = Number(m[1]), mo = Number(m[2]);
    if (d >= 1 && d <= 31 && mo >= 1 && mo <= 12) {
      return `${m[3]}-${String(mo).padStart(2, '0')}-${String(d).padStart(2, '0')}`;
    }
  }
  return '';
}

function bump(map, key) {
  if (!key) return;
  map.set(key, (map.get(key) || 0) + 1);
}

function topList(map, limit) {
  return [...map.entries()]
    .sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]))
    .slice(0, limit == null ? Infinity : limit)
    .map(([name, count]) => ({ name, count }));
}

// Nationality/citizenship where the authority states it, else the address
// country. Counted once per entity so a party with five addresses in the same
// country does not outweigh five separate parties. Labels are normalized across
// authorities (see lib/countries) — otherwise Russia lands in four rows.
function countriesOf(e) {
  const out = new Set();
  for (const a of e.attributes || []) {
    if (a.label === 'Nationality' || a.label === 'Citizenship') {
      for (const c of parseCountries(a.value)) out.add(c);
    }
  }
  if (!out.size) {
    for (const ad of e.addresses || []) {
      const c = canonicalCountry(ad.country);
      if (c) out.add(c);
    }
  }
  return out;
}

const OWNERSHIP_RE = /own|control|subsid|parent|beneficial/i;

function compute(snapshot) {
  const entities = snapshot.entities || [];
  const now = Date.now();

  const byAuthority = new Map(), byList = new Map(), byType = new Map();
  const byMeasure = new Map(), byProgram = new Map(), byCountry = new Map();
  const byYear = new Map(), byMonth = new Map();
  // Per-authority timeline of the trailing months, for the stacked view.
  const monthByAuthority = new Map();

  let dated = 0, withAliases = 0, withIdentifiers = 0, withRelationships = 0;
  let relationshipEdges = 0, ownershipEdges = 0;
  let added30 = 0, added90 = 0, added365 = 0;
  let newest = '', oldest = '';

  const recent = [];

  for (const e of entities) {
    const authority = e.authority || 'OFAC';
    bump(byAuthority, authority);
    bump(byList, e.list);
    bump(byType, e.type || 'Unknown');
    for (const m of e.sanctionsTypes && e.sanctionsTypes.length ? e.sanctionsTypes : (e.sanctionsType ? [e.sanctionsType] : [])) bump(byMeasure, m);
    for (const p of e.programs || []) bump(byProgram, p);
    for (const c of countriesOf(e)) bump(byCountry, c);

    if ((e.names || []).length > 1) withAliases++;
    if ((e.identifiers || []).length || (e.idDocuments || []).length) withIdentifiers++;
    const rels = e.relationships || [];
    if (rels.length) withRelationships++;
    relationshipEdges += rels.length;
    for (const r of rels) if (OWNERSHIP_RE.test(r.type || '')) ownershipEdges++;

    const date = normalizeDate(e.datePublished);
    if (!date) continue;
    dated++;
    if (!newest || date > newest) newest = date;
    if (!oldest || date < oldest) oldest = date;

    bump(byYear, date.slice(0, 4));
    const month = date.slice(0, 7);
    bump(byMonth, month);
    if (!monthByAuthority.has(authority)) monthByAuthority.set(authority, new Map());
    bump(monthByAuthority.get(authority), month);

    const age = now - Date.parse(date + 'T00:00:00Z');
    if (age >= 0) {
      if (age <= 30 * DAY) added30++;
      if (age <= 90 * DAY) added90++;
      if (age <= 365 * DAY) added365++;
    }

    recent.push({
      id: e.id, name: e.name, authority, list: e.list, type: e.type || 'Unknown',
      title: e.title || '', programs: (e.programs || []).slice(0, 6),
      measures: e.sanctionsTypes || (e.sanctionsType ? [e.sanctionsType] : []),
      date,
      aliases: Math.max(0, (e.names || []).length - 1),
      country: [...countriesOf(e)][0] || '',
    });
  }

  recent.sort((a, b) => (a.date < b.date ? 1 : a.date > b.date ? -1 : a.name.localeCompare(b.name)));

  // Global head + each authority's own head. OFAC publishes far more often than
  // the others, so a purely global window would leave the EU/UN/UK/BIS filters
  // showing nothing.
  const picked = new Set(recent.slice(0, RECENT_LIMIT).map((r) => r.id));
  const perAuthority = new Map();
  for (const r of recent) {
    const n = perAuthority.get(r.authority) || 0;
    if (n >= RECENT_PER_AUTHORITY) continue;
    perAuthority.set(r.authority, n + 1);
    picked.add(r.id);
  }
  const recentOut = recent.filter((r) => picked.has(r.id));

  // Trailing window ending at the newest designation in the snapshot, so a list
  // that has not published for a few days does not render as a gap at the edge.
  const endMonth = newest ? newest.slice(0, 7) : new Date(now).toISOString().slice(0, 7);
  const months = [];
  let [ey, em] = endMonth.split('-').map(Number);
  for (let i = 0; i < TIMELINE_MONTHS; i++) {
    const key = `${ey}-${String(em).padStart(2, '0')}`;
    months.unshift(key);
    if (--em === 0) { em = 12; ey--; }
  }
  const authorities = topList(byAuthority, null).map((a) => a.name);

  const endYear = Number(endMonth.slice(0, 4));
  const years = [];
  for (let y = endYear - TIMELINE_YEARS + 1; y <= endYear; y++) years.push(String(y));

  return {
    generatedAt: new Date().toISOString(),
    snapshot: {
      source: snapshot.source, isLive: !!snapshot.isLive,
      publicationId: snapshot.publicationId, publishedDate: snapshot.publishedDate || '',
      retrievedAt: snapshot.retrievedAt,
    },
    totals: {
      entities: entities.length,
      authorities: byAuthority.size,
      lists: byList.size,
      programs: byProgram.size,
      countries: byCountry.size,
      dated,
      undated: entities.length - dated,
      withAliases, withIdentifiers, withRelationships,
      relationshipEdges, ownershipEdges,
      added30, added90, added365,
      newest, oldest,
    },
    byAuthority: topList(byAuthority, null),
    byList: topList(byList, TOP_N),
    byType: topList(byType, null),
    byMeasure: topList(byMeasure, TOP_N),
    byProgram: topList(byProgram, TOP_N),
    byCountry: topList(byCountry, TOP_N),
    timeline: {
      years: years.map((y) => ({ label: y, count: byYear.get(y) || 0 })),
      months: months.map((mk) => ({
        label: mk,
        count: byMonth.get(mk) || 0,
        byAuthority: Object.fromEntries(authorities.map((a) => [a, (monthByAuthority.get(a) || new Map()).get(mk) || 0])),
      })),
      authorities,
    },
    recent: recentOut,
  };
}

let cachedFor = null, cached = null;
function snapshotStats(snapshot) {
  if (cachedFor !== snapshot) {
    const t = Date.now();
    cached = compute(snapshot);
    cachedFor = snapshot;
    console.log(`Stats computed: ${snapshot.count} entities in ${Date.now() - t}ms`);
  }
  return cached;
}

module.exports = { snapshotStats, normalizeDate, RECENT_LIMIT, RECENT_PER_AUTHORITY };

File: ios/SanctinelCore/Sources/SanctinelCore/Stats.swift
import Foundation

/*
 * Snapshot analytics — a port of lib/stats.js.
 *
 * Everything here is derived from the SAME snapshot the screening path uses, so
 * a statistic and a search result can never disagree. On the phone this runs
 * once, when a snapshot is imported, and the result is stored: a full pass over
 * 38k entities is not something to repeat on every tab switch.
 *
 * Dates: each authority publishes its listing date in its own format — OFAC, EU,
 * UN and BIS in ISO, UK OFSI in DD/MM/YYYY. They are normalized to ISO here so
 * the timeline and the "recent" ranking are comparable. The date is the
 * AUTHORITY'S designation date, not when the app fetched it.
 */

public struct SnapshotStats: Codable, Sendable {
    public struct Totals: Codable, Sendable {
        public var entities = 0
        public var authorities = 0
        public var lists = 0
        public var programs = 0
        public var countries = 0
        public var dated = 0
        public var undated = 0
        public var withAliases = 0
        public var withIdentifiers = 0
        public var withRelationships = 0
        public var relationshipEdges = 0
        public var ownershipEdges = 0
        public var added30 = 0
        public var added90 = 0
        public var added365 = 0
        public var newest = ""
        public var oldest = ""
        public init() {}
    }

    public struct Bucket: Codable, Sendable, Hashable, Identifiable {
        public let name: String
        public let count: Int
        public var id: String { name }
        public init(name: String, count: Int) { self.name = name; self.count = count }
    }

    public struct Year: Codable, Sendable, Hashable, Identifiable {
        public let label: String
        public let count: Int
        public var id: String { label }
    }

    public struct RecentDesignation: Codable, Sendable, Hashable, Identifiable {
        public let id: String
        public let name: String
        public let authority: String
        public let list: String
        public let type: String
        public let title: String
        public let programs: [String]
        public let measures: [String]
        public let date: String
        public let aliases: Int
        public let country: String
    }

    public var generatedAt = ""
    public var totals = Totals()
    public var byAuthority: [Bucket] = []
    public var byList: [Bucket] = []
    public var byType: [Bucket] = []
    public var byMeasure: [Bucket] = []
    public var byProgram: [Bucket] = []
    public var byCountry: [Bucket] = []
    public var years: [Year] = []
    public var recent: [RecentDesignation] = []

    public init() {}
}

public enum StatsBuilder {
    static let recentLimit = 150         // most recent designations across all lists
    static let recentPerAuthority = 40   // …plus this many per authority, so the
                                         // authority filter is never empty for a
                                         // list that has not published lately
    static let topN = 15
    static let timelineYears = 16
    private static let day: TimeInterval = 86_400

    /// A date only counts if it exists on the calendar: "31/02/2020" would
    /// otherwise sort into the timeline and render blank.
    static func isRealDate(_ y: Int, _ m: Int, _ d: Int) -> Bool {
        guard m >= 1, m <= 12, d >= 1 else { return false }
        var components = DateComponents()
        components.year = y; components.month = m; components.day = d
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        guard let date = calendar.date(from: components) else { return false }
        let back = calendar.dateComponents([.year, .month, .day], from: date)
        return back.year == y && back.month == m && back.day == d
    }

    /// Accepts "YYYY-MM-DD" / ISO timestamps and "D/M/YYYY" (UK OFSI).
    public static func normalizeDate(_ value: String) -> String {
        let s = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.isEmpty { return "" }
        let chars = Array(s)

        // ^(\d{4})-(\d{2})-(\d{2})
        if chars.count >= 10,
           let y = Int(String(chars[0..<4])), chars[4] == "-",
           let m = Int(String(chars[5..<7])), chars[7] == "-",
           let d = Int(String(chars[8..<10])),
           chars[0..<4].allSatisfy(\.isNumber), chars[5..<7].allSatisfy(\.isNumber),
           chars[8..<10].allSatisfy(\.isNumber) {
            return isRealDate(y, m, d) ? String(chars[0..<10]) : ""
        }

        // ^(\d{1,2})/(\d{1,2})/(\d{4})$
        let parts = s.split(separator: "/", omittingEmptySubsequences: false)
        if parts.count == 3,
           parts[0].count >= 1, parts[0].count <= 2, parts[0].allSatisfy(\.isNumber),
           parts[1].count >= 1, parts[1].count <= 2, parts[1].allSatisfy(\.isNumber),
           parts[2].count == 4, parts[2].allSatisfy(\.isNumber),
           let d = Int(parts[0]), let m = Int(parts[1]), let y = Int(parts[2]),
           isRealDate(y, m, d) {
            return String(format: "%04d-%02d-%02d", y, m, d)
        }
        return ""
    }

    /// Nationality/citizenship where the authority states it, else the address
    /// country. Counted once per entity so a party with five addresses in one
    /// country does not outweigh five separate parties.
    static func countries(of entity: [String: Any]) -> [String] {
        var out: [String] = []
        for a in entity["attributes"] as? [[String: Any]] ?? [] {
            let label = RecordBuilder.string(a["label"])
            // By canonical kind: matching the two literal strings meant OFAC's own
            // wording — `Nationality Country` (5,821 attributes) and `Citizenship
            // Country` (1,142) — never matched, so for OFAC parties this fell
            // through to the address country and the chart described where a party
            // is located rather than what it is a national of.
            let kind = Vocab.attributeKind(label)
            guard kind == "nationality" || kind == "citizenship" else { continue }
            for c in Countries.parse(RecordBuilder.string(a["value"])) where !out.contains(c) {
                out.append(c)
            }
        }
        if out.isEmpty {
            for ad in entity["addresses"] as? [[String: Any]] ?? [] {
                let c = Countries.canonical(RecordBuilder.string(ad["country"]))
                if !c.isEmpty, !out.contains(c) { out.append(c) }
            }
        }
        return out
    }

    public static func compute(entities: [[String: Any]], now: Date = Date()) -> SnapshotStats {
        var byAuthority: [String: Int] = [:], byList: [String: Int] = [:], byType: [String: Int] = [:]
        var byMeasure: [String: Int] = [:], byProgram: [String: Int] = [:], byCountry: [String: Int] = [:]
        var byYear: [String: Int] = [:]

        var totals = SnapshotStats.Totals()
        var recent: [SnapshotStats.RecentDesignation] = []
        let nowSeconds = now.timeIntervalSince1970

        func bump(_ map: inout [String: Int], _ key: String) {
            if key.isEmpty { return }
            map[key, default: 0] += 1
        }

        for entity in entities {
            let authorityRaw = RecordBuilder.string(entity["authority"])
            let authority = authorityRaw.isEmpty ? "OFAC" : authorityRaw
            bump(&byAuthority, authority)
            bump(&byList, RecordBuilder.string(entity["list"]))
            let type = RecordBuilder.string(entity["type"])
            bump(&byType, type.isEmpty ? "Unknown" : type)

            let measures = measuresOf(entity)
            for m in measures { bump(&byMeasure, m) }
            for p in entity["programs"] as? [String] ?? [] { bump(&byProgram, p) }

            let countryList = countries(of: entity)
            for c in countryList { bump(&byCountry, c) }

            let names = entity["names"] as? [[String: Any]] ?? []
            if names.count > 1 { totals.withAliases += 1 }
            let identifiers = entity["identifiers"] as? [[String: Any]] ?? []
            let documents = entity["idDocuments"] as? [[String: Any]] ?? []
            if !identifiers.isEmpty || !documents.isEmpty { totals.withIdentifiers += 1 }

            let relationships = entity["relationships"] as? [[String: Any]] ?? []
            if !relationships.isEmpty { totals.withRelationships += 1 }
            totals.relationshipEdges += relationships.count
            for r in relationships where Ownership.isOwnership(RecordBuilder.string(r["type"])) {
                totals.ownershipEdges += 1
            }

            let date = normalizeDate(RecordBuilder.string(entity["datePublished"]))
            if date.isEmpty { continue }
            totals.dated += 1
            if totals.newest.isEmpty || date > totals.newest { totals.newest = date }
            if totals.oldest.isEmpty || date < totals.oldest { totals.oldest = date }
            bump(&byYear, String(date.prefix(4)))

            if let parsed = utcMidnight(date) {
                let age = nowSeconds - parsed
                if age >= 0 {
                    if age <= 30 * day { totals.added30 += 1 }
                    if age <= 90 * day { totals.added90 += 1 }
                    if age <= 365 * day { totals.added365 += 1 }
                }
            }

            recent.append(SnapshotStats.RecentDesignation(
                id: RecordBuilder.string(entity["id"]),
                name: RecordBuilder.string(entity["name"]),
                authority: authority,
                list: RecordBuilder.string(entity["list"]),
                type: type.isEmpty ? "Unknown" : type,
                title: RecordBuilder.string(entity["title"]),
                programs: Array((entity["programs"] as? [String] ?? []).prefix(6)),
                measures: measures,
                date: date,
                aliases: max(0, names.count - 1),
                country: countryList.first ?? ""
            ))
        }

        recent.sort { a, b in
            if a.date != b.date { return a.date > b.date }
            return compareLocalized(a.name, b.name) == .orderedAscending
        }

        // Global head plus each authority's own head. OFAC publishes far more
        // often than the others, so a purely global window would leave the
        // EU/UN/UK/BIS filters showing nothing.
        var picked = Set(recent.prefix(recentLimit).map(\.id))
        var perAuthority: [String: Int] = [:]
        for r in recent {
            let n = perAuthority[r.authority] ?? 0
            if n >= recentPerAuthority { continue }
            perAuthority[r.authority] = n + 1
            picked.insert(r.id)
        }
        let recentOut = recent.filter { picked.contains($0.id) }

        totals.entities = entities.count
        totals.undated = entities.count - totals.dated
        totals.authorities = byAuthority.count
        totals.lists = byList.count
        totals.programs = byProgram.count
        totals.countries = byCountry.count

        // Trailing window ending at the newest designation in the snapshot, so a
        // list that has not published for a few days is not a gap at the edge.
        let endYear = Int(totals.newest.prefix(4)) ?? Calendar(identifier: .gregorian)
            .component(.year, from: now)
        var years: [SnapshotStats.Year] = []
        for y in (endYear - timelineYears + 1)...endYear {
            let label = String(y)
            years.append(SnapshotStats.Year(label: label, count: byYear[label] ?? 0))
        }

        var stats = SnapshotStats()
        stats.generatedAt = ISO8601DateFormatter().string(from: now)
        stats.totals = totals
        stats.byAuthority = topList(byAuthority, limit: nil)
        stats.byList = topList(byList, limit: topN)
        stats.byType = topList(byType, limit: nil)
        stats.byMeasure = topList(byMeasure, limit: topN)
        stats.byProgram = topList(byProgram, limit: topN)
        stats.byCountry = topList(byCountry, limit: topN)
        stats.years = years
        stats.recent = recentOut
        return stats
    }

    static func measuresOf(_ entity: [String: Any]) -> [String] {
        if let list = entity["sanctionsTypes"] as? [String], !list.isEmpty { return list }
        let single = RecordBuilder.string(entity["sanctionsType"])
        return single.isEmpty ? [] : [single]
    }

    static func topList(_ map: [String: Int], limit: Int?) -> [SnapshotStats.Bucket] {
        var entries = map.map { SnapshotStats.Bucket(name: $0.key, count: $0.value) }
        entries.sort { a, b in
            a.count == b.count ? compareLocalized(a.name, b.name) == .orderedAscending : a.count > b.count
        }
        if let limit { entries = Array(entries.prefix(limit)) }
        return entries
    }

    /// `String.prototype.localeCompare` — ICU collation, so "a" sorts before "B".
    /// Pinned to en_US rather than the device locale so the ordering is the same
    /// on every phone and the same as the server's.
    static func compareLocalized(_ a: String, _ b: String) -> ComparisonResult {
        a.compare(b, options: [], range: nil, locale: Locale(identifier: "en_US"))
    }

    static func utcMidnight(_ isoDay: String) -> TimeInterval? {
        var components = DateComponents()
        components.year = Int(isoDay.prefix(4))
        components.month = Int(isoDay.dropFirst(5).prefix(2))
        components.day = Int(isoDay.dropFirst(8).prefix(2))
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        return calendar.date(from: components)?.timeIntervalSince1970
    }
}
